app/scripts/lib/multichain-api/multichainSubscriptionManager.ts

```diff
@@ -0,0 +1,153 @@
+import {
+  NetworkControllerFindNetworkClientIdByChainIdAction,
+  NetworkControllerGetNetworkClientByIdAction,
+} from '@metamask/network-controller';
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { parseCaipChainId } from '@metamask/utils';
+import { toHex } from '@metamask/controller-utils';
+import { Scope } from './scope';
+
+type SubscriptionManager = {
+  events: {
+    on: (
+      event: string,
+      listener: MultichainSubscriptionManager['onNotification'],
+    ) => void;
+    off: (
+      event: string,
+      listener: MultichainSubscriptionManager['onNotification'],
+    ) => void;
+  };
+  destroy?: () => void;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
+const createSubscriptionManager = require('@metamask/eth-json-rpc-filters/subscriptionManager');
+
+type MultichainSubscriptionManagerOptions = {
+  findNetworkClientIdByChainId: NetworkControllerFindNetworkClientIdByChainIdAction['handler'];
+  getNetworkClientById: NetworkControllerGetNetworkClientByIdAction['handler'];
+};
+
+export default class MultichainSubscriptionManager extends SafeEventEmitter {
+  private subscriptionsByChain: {
+    [scope: string]: {
+      [domain: string]: MultichainSubscriptionManager['onNotification'];
+    };
+  };
+
+  private findNetworkClientIdByChainId: NetworkControllerFindNetworkClientIdByChainIdAction['handler'];
+
+  private getNetworkClientById: NetworkControllerGetNetworkClientByIdAction['handler'];
+
+  private subscriptionManagerByChain: { [scope: string]: SubscriptionManager };
+
+  private subscriptionsCountByScope: { [scope: string]: number };
+
+  constructor(options: MultichainSubscriptionManagerOptions) {
+    super();
+    this.findNetworkClientIdByChainId = options.findNetworkClientIdByChainId;
+    this.getNetworkClientById = options.getNetworkClientById;
+    this.subscriptionManagerByChain = {};
+    this.subscriptionsByChain = {};
+    this.subscriptionsCountByScope = {};
+  }
+
+  onNotification(scope: Scope, domain: string, message: unknown) {
+    this.emit('notification', domain, {
+      method: 'wallet_invokeMethod',
+      params: {
+        scope,
+        request: message,
+      },
+    });
+  }
+
+  subscribe(scope: Scope, domain: string) {
+    this.subscriptionsCountByScope[scope] =
+      this.subscriptionsCountByScope[scope] || 0;
+    this.subscriptionsCountByScope[scope] += 1;
```

Comment:
could this get out of sync is something fails below? Should this go at the bottom of the method?